File: backend/src/services/auth_service.go
package services

import (
	"errors"

	"live-auction-system/backend/src/models"
	"live-auction-system/backend/src/repositories"
	"live-auction-system/backend/src/utils"
)

var (
	ErrInvalidCredentials = errors.New("invalid credentials")
	ErrUserExists         = errors.New("user already exists")
)

type AuthService struct {
	userRepo *repositories.UserRepository
}

func NewAuthService() *AuthService {
	return &AuthService{
		userRepo: repositories.NewUserRepository(),
	}
}

func (s *AuthService) Register(req *models.RegisterRequest) (*models.AuthResponse, error) {
	existingUser, _ := s.userRepo.FindByUsername(req.Username)
	if existingUser != nil {
		return nil, ErrUserExists
	}

	existingEmail, _ := s.userRepo.FindByEmail(req.Email)
	if existingEmail != nil {
		return nil, ErrUserExists
	}

	role := models.RoleUser
	if req.Role == "seller" {
		role = models.RoleSeller
	}

	user := &models.User{
		Username: req.Username,
		Password: req.Password,
		Email:    req.Email,
		Role:     role,
	}

	if err := s.userRepo.Create(user); err != nil {
		return nil, err
	}

	token, err := utils.GenerateToken(user.ID, user.Username, string(user.Role))
	if err != nil {
		return nil, err
	}

	return &models.AuthResponse{
		Token: token,
		User:  *user,
	}, nil
}

func (s *AuthService) Login(req *models.LoginRequest) (*models.AuthResponse, error) {
	user, err := s.userRepo.FindByUsername(req.Username)
	if err != nil {
		return nil, ErrInvalidCredentials
	}

	if !s.userRepo.CheckPassword(user, req.Password) {
		return nil, ErrInvalidCredentials
	}

	token, err := utils.GenerateToken(user.ID, user.Username, string(user.Role))
	if err != nil {
		return nil, err
	}

	return &models.AuthResponse{
		Token: token,
		User:  *user,
	}, nil
}

func (s *AuthService) GetUserByID(id int64) (*models.User, error) {
	return s.userRepo.FindByID(id)
}

File: backend/src/services/auth.service.ts
import bcrypt from 'bcrypt';
import jwt from 'jsonwebtoken';
import { prisma } from '../lib/prisma';
import config from '../config';

export class AuthService {
  async register(phone: string, password: string, nickname?: string) {
    const hashedPassword = await bcrypt.hash(password, 10);
    
    const user = await prisma.user.create({
      data: {
        phone,
        password: hashedPassword,
        nickname: nickname || phone
      }
    });
    
    const token = this.generateToken(user.id, user.phone);
    
    return {
      user: {
        id: user.id,
        phone: user.phone,
        nickname: user.nickname
      },
      token
    };
  }
  
  async login(phone: string, password: string) {
    const user = await prisma.user.findUnique({
      where: { phone }
    });
    
    if (!user) {
      throw new Error('用户不存在');
    }
    
    const isPasswordValid = await bcrypt.compare(password, user.password);
    
    if (!isPasswordValid) {
      throw new Error('密码错误');
    }
    
    const token = this.generateToken(user.id, user.phone);
    
    return {
      user: {
        id: user.id,
        phone: user.phone,
        nickname: user.nickname
      },
      token
    };
  }
  
  private generateToken(userId: string, phone: string) {
    return jwt.sign(
      { id: userId, phone },
      config.jwt.secret,
      { expiresIn: config.jwt.expiresIn }
    );
  }
}
